### ISSUE TITLE
Flaky `Missing output: my-package (used in pin_subpackage)` error

### ISSUE THREAD

**Issue by pavelzw** [prefix-dev/rattler-build#553]
Let's take a look at this recipe:

```yml
# yaml-language-server: $schema=https://raw.githubusercontent.com/prefix-dev/recipe-format/main/schema.json

context:
  name: my-package
  version: '0.1.0'

outputs:
  - package:
      name: ${{ name }}
      version: ${{ version }}
    build:
      noarch: generic

  - package:
      name: ${{ name | replace('-', '_') }}
      version: ${{ version }}
    build:
      noarch: generic
    requirements:
      run:
        - ${{ pin_subpackage(name, exact=true) }}
```

Often, the output is this as expected:

```
❯ rattler-build build -r r/recipe.yaml
No target platform specified, using current platform
Found variants:

my-package-0.1.0-h4616a5c_0
╭─────────────────┬─────────╮
│ Variant         ┆ Version │
╞═════════════════╪═════════╡
│ target_platform ┆ noarch  │
╰─────────────────┴─────────╯

my_package-0.1.0-h831704a_0
╭─────────────────┬──────────────────╮
│ Variant         ┆ Version          │
╞═════════════════╪══════════════════╡
│ my-package      ┆ 0.1.0 h4616a5c_0 │
│ target_platform ┆ noarch           │
╰─────────────────┴──────────────────╯

Work dir: "/private/tmp/rattler-test/output/bld/rattler-build_my-package_1706476231/work"
Build script: "/private/tmp/rattler-test/output/bld/rattler-build_my-package_1706476231/work/conda_build.sh"
Copying done!
Relink done!

Files in package:

  - info/about.json
  - info/hash_input.json
  - info/index.json
  - info/paths.json
  - info/recipe/recipe.yaml
  - info/recipe/rendered_recipe.yaml
  - info/recipe/variant_config.yaml
Creating target folder "/private/tmp/rattler-test/output/noarch"

Output: my-package-0.1.0-h4616a5c_0

Variant configuration:
┌─────────────────┬─────────┐
│ Variant         ┆ Version │
╞═════════════════╪═════════╡
│ target_platform ┆ noarch  │
└─────────────────┴─────────┘




Running tests
Removing previously cached package "/Users/pavel/Library/Caches/rattler/cache/pkgs/my-package-0.1.0-h4616a5c_0"
Creating test environment in "/private/tmp/rattler-test/output/bld/rattler-build_my-package_1706476231/work/test"
Collecting tests from "/Users/pavel/Library/Caches/rattler/cache/pkgs/my-package-0.1.0-h4616a5c_0"
Work dir: "/private/tmp/rattler-test/output/bld/rattler-build_my_package_1706476231/work"
Build script: "/private/tmp/rattler-test/output/bld/rattler-build_my_package_1706476231/work/conda_build.sh"
Copying done!
Relink done!

Files in package:

  - info/about.json
  - info/hash_input.json
  - info/index.json
  - info/paths.json
  - info/recipe/recipe.yaml
  - info/recipe/rendered_recipe.yaml
  - info/recipe/variant_config.yaml
Creating target folder "/private/tmp/rattler-test/output/noarch"

Output: my_package-0.1.0-h831704a_0

Variant configuration:
┌─────────────────┬──────────────────┐
│ Variant         ┆ Version          │
╞═════════════════╪══════════════════╡
│ my-package      ┆ 0.1.0 h4616a5c_0 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ target_platform ┆ noarch           │
└─────────────────┴──────────────────┘

Run dependencies:
╭────────────┬─────────────────────────╮
│ Name       ┆ Spec                    │
╞════════════╪═════════════════════════╡
│ my-package ┆ ==0.1.0 h4616a5c_0 (PS) │
╰────────────┴─────────────────────────╯




Running tests
Removing previously cached package "/Users/pavel/Library/Caches/rattler/cache/pkgs/my_package-0.1.0-h831704a_0"
Creating test environment in "/private/tmp/rattler-test/output/bld/rattler-build_my_package_1706476231/work/test"
Collecting tests from "/Users/pavel/Library/Caches/rattler/cache/pkgs/my_package-0.1.0-h831704a_0"
```

But sometimes, I get this error:

```
❯ rattler-build build -r r/recipe.yaml
No target platform specified, using current platform
Error:   × Missing output: my-package (used in pin_subpackage)
```

When I replace `${{ pin_subpackage(name, exact=true) }}` with `${{ pin_subpackage('my-package', exact=true) }}`, I cannot replicate the error.

I unfortunately couldn't find other patterns to reproduce this...

**Comment by wolfv** [comment]
Interesting, sounds like we're missing an edge in the graph for the topological sorting of the outputs. I'll write some tests tomorrow.

**Comment by wolfv** [comment]
Ok, I briefly debugged a bit – the issue is that we are operating on the `AST` to find the used variables from the recipe. In this case, we determine that `name` is a used variable (correct), but we don't resolve it from the context (so the edge is not correctly introduced between the packages and the order isn't right).

I'll see if I can think of better ways to do this ...

**Comment by wolfv** [comment]
The other part of the issue is that we currently don't take `run` dependencies into account for computing the graph – but later, when executing the tests, run dependencies are obviously used. 

I am not sure though wether adding run dependencies is the right thing to do ... but maybe there should never be a situation where one has cyclic graphs due to run dependencies (it would be valid though, otherwise).

**Comment by wolfv** [comment]
This should be fixed in #617 :)
